storagemarket/deal_execution.go

```diff
@@ -80,21 +63,32 @@ func (p *Provider) execDeal(ctx context.Context, pub event.Emitter, deal *types.
 
 	// Transfer Data
 	if deal.Checkpoint < dealcheckpoints.Transferred {
-		if err := p.transferAndVerify(ctx, pub, deal); err != nil {
+		if err := p.transferAndVerify(dh.transferCtx, pub, deal); err != nil {
+			dh.transferWriteOnce.Do(func() {
+				dh.transferredCancelled <- transferCancelResp{}
+				close(dh.transferredCancelled)
+			})
```

Comment:
This code operates on variables that are internal to the deal handler struct, so I suggest putting it in a method on the dealHandler struct. Something like `dh.transferComplete()`